File: Week_06/G20200343040030/edit-distance.js
/**
 * @param {string} word1
 * @param {string} word2
 * @return {number}
 */
let minDistance = (word1, word2) => {
  //1.初始化
  let n = word1.length,
    m = word2.length;
  let dp = new Array(n + 1).fill(0).map(() => new Array(m + 1).fill(0));
  for (let i = 0; i <= n; i++) {
    dp[i][0] = i;
  }
  for (let j = 0; j <= m; j++) {
    dp[0][j] = j;
  }
  //2.dp
  for (let i = 0; i <= n; i++) {
    for (let j = 0; j <= m; j++) {
      if (i * j) {
        dp[i][j] =
          word1[i - 1] == word2[j - 1]
            ? dp[i - 1][j - 1]
            : Math.min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1]) + 1;
      } else {
        dp[i][j] = i + j;
      }
    }
  }
  return dp[n][m];
};

File: Week_06/G20200343040044/edit-distance.java
package 第五周;

class Solution {
    public int minDistance(String word1, String word2) {
        int[][] dp = new int[word1.length()+1][word2.length()+1];
        for(int i = 0;i<=word1.length();i++) dp[i][0] = i;
        for(int j = 0;j<=word2.length();j++) dp[0][j] = j;
        for(int i = 0;i<word1.length();i++){
            for(int j = 0;j<word2.length();j++){
                if(word1.charAt(i) == word2.charAt(j)) dp[i+1][j+1] = dp[i][j];
                else dp[i+1][j+1] = Math.min(dp[i][j],Math.min(dp[i+1][j],dp[i][j+1])) + 1;
            }
        }
        return dp[word1.length()][word2.length()];
    }
}